CONTRIBUTING.md

```diff
@@ -45,9 +45,88 @@ include your changes.
   the name of the thing you're changing in at the beginning of the message,
   followed by a colon: the plugin name for plugins, "docs" for documentation
   files, "coretasks" for `coretasks.py`, "db" for the database feature, etc.
-* Python files should always have  `from __future__ import generator_stop`
+* Python files should always have `from __future__ import annotations`
   as the first line after the module docstring.
 
+Running from source
+-------------------
+
+Sopel is your typical Python project: you can install it from source, edit
+the code using standard tooling, and take advantage of your favorite code
+editor.
+
+Assuming you are using standard tools:
+
+* Create a virtualenv.
+  * You can use `virtualenv` directly, `virtualenvwrapper`, or any tool that
+    allows you to create and manage your virtualenv.
+* Fork Sopel's repository, and clone your fork locally.
+* Activate your virtualenv, and `cd` into your clone's folder.
+* Ensure you have the latest version of `pip`, and install `wheel`.
+* Install Sopel from source as an editable install.
+  * If you don't use a project manager (like Poetry or Hatch), you can do
+    that with `pip install -e .` from within the clone's directory, where
+    there is a `pyproject.toml` file.
+* Install development tools used to run tests and linters; there is a specific
+  requirement file you can use with `pip install -U -r dev-requirements.txt`.
+* Run `make qa` to run linters and tests.
+* Run `make cleandoc` or just `make docs` to build the documentation locally.
+
+At this point, you are all set, at least for the Python environment; all
+that's left is for you to configure your code editor the way you like, and read and
+write some code and documentation.
+
+Using branches
+--------------
+
+As previously stated, you should fork Sopel's repository to implement your
+changes and commit them. Moreover, we advise you to work from your own branch.
+
+To setup your local clone, we suggest the following steps:
+
+```
+$ cd path/to/your/dev/folder
+$ mkdir sopel
+$ cd sopel/
+$ git clone git@github.com:<USERNAME>/sopel.git .
+$ git remote add upstream git@github.com:sopel-irc/sopel.git
+$ git checkout -b <your-branch-name>
+```
+
+With that workflow, you have your local clone with a link to the upstream
+repository, and a branch to work on. Once you are done with your work, you can
+commit, and push to your repository.
+
+You'll probably need, at some point, to update your repository with the new
+commits from upstream. We suggest the following steps:
+
+```
+$ git checkout master
+$ git fetch upstream
+$ git rebase upstream/master
```

Comment:
Why not `git merge --ff-only upstream/master`? Unless I'm way off base about how rebase works, using it here may change SHAs on the downstream fork because `pick`ing a commit _copies_ it (and applies a new committer date). I've never seen someone suggest using `rebase` for this task before. 🤔

```suggestion
$ git merge --ff-only upstream/master
```

True, a straight merge risks adding a merge commit, but `--ff-only` aborts if the merge isn't a fast-forward.

Then again, if the user truly hasn't touched their own master branch, the nuclear option is fine too 😁:

```suggestion
$ git reset --hard upstream/master
```

Reply:
Ha ah. I do `git rebase upstream/master` all the time, because if you don't commit on your master branch, then what rebase does it take all the commit from upstream/master, and apply all the commits that exists only in your version. Therefore, only *your* commits are changed.

Basically, it's like `merge --ff-only`.

Reply:
That'll count as "way off base about how rebase works", then. 😅